dipy/tracking/local/localtrack.pyx

```diff
@@ -208,3 +199,303 @@ cdef int _local_tracker(
+            # no valid diffusion direction to follow
+            tissue_class[0] = INVALIDPOINT
+        else:
+            for j in range(3):
+                voxdir[j] = dir[j] / voxel_size[j]
+
+            fixed_step(point, voxdir, step_size)
+            copypoint(point, &streamline[i, 0])
+            copypoint(dir, &directions[i, 0])
+            tissue_class[0] = tc.check_point_c(point)
+            i += 1
+        if tissue_class[0] == TRACKPOINT:
+            # The tracking continues normally
+            continue
+        elif tissue_class[0] == INVALIDPOINT:
+            if pft_trial < pft_max_trials and i > 1:
+                back_steps = min(i - 1, pft_max_nbr_back_steps)
+                front_steps = min(strl_array_len - i - back_steps - 1,
+                                  pft_max_nbr_front_steps)
+                front_steps = max(0, front_steps)
+                i = _pft(streamline, i - back_steps, directions, dg, tc,
+                         voxel_size, step_size, tissue_class,
+                         back_steps + front_steps, particle_count,
+                         particle_paths, particle_dirs, particle_weights,
+                         particle_states)
+                pft_trial += 1
+                # update the current point with the PFT results
+                copypoint(&streamline[i-1, 0], point)
+                copypoint(&directions[i-1, 0], dir)
+
+                if tissue_class[0] != TRACKPOINT:
+                    # The tracking stops. PFT returned a valid stopping point
+                    # (ENDPOINT, OUTSIDEIMAGE) or failed to find one
+                    # (INVALIDPOINT, PYERROR)
+                    break
+            else:
+                # PFT was run more times than `pft_max_trials` without finding
+                # a valid stopping point. The tracking stops with INVALIDPOINT.
+                break
+        else:
+            # The tracking stops with a valid point (ENDPOINT, OUTSIDEIMAGE)
+            # or an invalid point (PYERROR)
+            break
+
+    if tissue_class[0] == OUTSIDEIMAGE or tissue_class[0] == PYERROR:
+        i -= 1
+    return i
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef _pft(np.float_t[:, :] streamline,
+          int streamline_i,
+          np.float_t[:, :] directions,
+          DirectionGetter dg,
+          ConstrainedTissueClassifier tc,
+          double* voxel_size,
+          double step_size,
+          TissueClass * tissue_class,
+          int pft_nbr_steps,
+          int particle_count,
+          np.float_t[:, :, :, :] particle_paths,
+          np.float_t[:, :, :, :] particle_dirs,
+          np.float_t[:] particle_weights,
+          np.int_t[:, :, :] particle_states):
+    cdef:
+        double sum_weights, sum_squared, N_effective, rdm_sample
+        double point[3]
+        double dir[3]
+        double voxdir[3]
+        double eps = np.finfo(np.float).eps
```

Comment:
float64 should be fine as we always use `double` in `trilinear_interpolate4d`. It might be safer to put eps slightly higher as we do multiple additions in `trilinear_interpolate4d`.